### PR TITLE
JS: Model more this-nodes as JQuery/DOM objects.

### PR DESCRIPTION
**Pull request by erik-krogh**
Gets a TP in CVE-2019-8331

The first commit models the `this` node in: 

```JavaScript
$.fn.myFancyHidePlugin = function () {
  this.hide(400)
}
```

And the second models the `this` node in: 

```JavaScript
$(mySelector).each(function () {
  $(this).hide();
});
```

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 1]
```diff
@@ -496,6 +496,40 @@ module JQuery {
         hasUnderlyingType("jQuery")
       }
     }
+
+    /**
+     * Gets a node that is written to `$.fn[something]`.
+     * JQuery plugins are registered this way.
+     */
+    private DataFlow::Node getAFnWrite() {
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/jQuery.qll:504]: Except for the requirement of the known plugin name, this seems to do exactly what you need: https://github.com/Semmle/ql/blob/801ce89c67c08e56beb0ebf88042d5b83ca4f1db/javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll#L55

**Comment by asgerf** [comment]
> "this is a wrapped set, so it must be a set of DOM nodes"

Could you elaborate on this? Are you saying jQuery collections may contain something other than DOM nodes?

**Comment by esbena** [comment]
Oh yes, young kids may not remember that `$(...)` was the convenient wrapper for arrays back in the day...

```js
> typeof $([1])[0]
"number"
```

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/jQuery.qll]: It is close enough that I can reuse it.

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 2]
```diff
@@ -624,4 +611,27 @@ module JQuery {
       node = getArgument(0)
     }
   }
+
+  /**
+   * Holds for jQuery plugin definitions of the form `$.fn.<pluginName> = <plugin>`.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/jQuery.qll:616]:
```suggestion
   * Holds for jQuery plugin definitions of the form `$.fn.<pluginName> = <plugin>` or $.extend($.fn, {<pluginName>, <plugin>})`.
```

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 3]
```diff
@@ -497,24 +497,11 @@ module JQuery {
...
-        result = write.getRhs()
-      )
-    }
-
     /**
      * Gets a node that is backtracked from a node written to `$.fn[something]`.
      */
     private DataFlow::SourceNode writtenToJqueryFN(DataFlow::TypeBackTracker t) {
-      t.start() and result = getAFnWrite().getALocalSource()
+      t.start() and result = any(DataFlow::Node plugin | jQueryPluginDefinition(_, plugin)).getALocalSource()
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/jQuery.qll:504]: Sorry. I should have pointed you here instead: https://github.com/Semmle/ql/blob/36b28386f80dc8489f6bce4ad4dad82bed2b7f83/javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll#L88-L93

**Comment by erik-krogh** [comment]
I expanded the scope of this PR slightly. 

But now we additionally get a TP/TN on CVE-2016-10735.

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 4]
```diff
@@ -590,4 +599,64 @@ module JQuery {
       node = getArgument(0)
     }
   }
+
+  /**
+   * Holds for jQuery plugin definitions of the form `$.fn.<pluginName> = <plugin>` or $.extend($.fn, {<pluginName>, <plugin>})`.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/jQuery.qll:604]:
```suggestion
   * Holds for jQuery plugin definitions of the form `$.fn.<pluginName> = <plugin>` or `$.extend($.fn, {<pluginName>, <plugin>})`.
```

**Comment by erik-krogh** [comment]
An [evaluation of nightly using `Xss.ql`](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-max.northeurope.cloudapp.azure.com_1587674616643) looks good.  
(And [a redo with the security suite on the weird results](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-max.northeurope.cloudapp.azure.com_1587748421151)).

**Pull request merged**